### ISSUE TITLE
typed/rackunit: check-eq? broke its contract

### ISSUE THREAD

**Issue by bennn** [racket/typed-racket#54]
check-eq? and check-equal? etc get blamed when I try to compare structs.

```
#lang typed/racket/base

(require typed/rackunit)

(define-struct foo ([bar : Any]))
(define foo1 (make-foo 42))

(check-eq? foo1 foo1)
```

The error:
```
check-eq?: broke its contract
  Attempted to use a higher-order value passed as `Any` in untyped code: #<foo>
  in: the 1st argument of
      (->* (Any Any) (any/c) any/c)
  contract from: (interface for check-eq?)
  blaming: (interface for check-eq?)
   (assuming the contract is correct)
  at: <pkgs>/typed-racket-more/typed/rackunit/main.rkt:18.2
```

**Comment by lexi-lambda** [comment]
I don't think this is an issue with `check-eq?`, this is an issue with `eq?`. This errors, too.

```racket
#lang racket

(module untyped racket
  (provide my-eq?)
  (define (my-eq? a b)
    (eq? a b)))

(module typed typed/racket
  (require/typed
   (submod ".." untyped)
   [my-eq? (Any Any -> Boolean)])
  (struct foo ([x : Any]))
  (define s (foo 'bar))
  (my-eq? s s))

(require 'typed)
```

**Comment by lexi-lambda** [comment]
Looks like this was already discussed on the mailing list here:
http://lists.racket-lang.org/dev/archive/2013-May/012455.html

**Comment by lexi-lambda** [comment]
@bennn This works, but it fails the test because the instances are no longer `eq?`.

```racket
#lang typed/racket/base

(require typed/rackunit)

(struct foo ([bar : Any]) #:transparent)
(define foo1 (foo 42))

(check-eq? foo1 foo1)
```

Marking the struct `#:transparent` seems to stop the value from being considered higher-order, but the contract wrapping the `Any` value means they are no longer the same object. Replacing `check-eq?` with `check-equal?` makes the test pass.

**Comment by bennn** [comment]
Cool, thanks for the mailing list link.

But still, I think this is an issue with `check-eq?`because library code should not be raising contract errors.

**Comment by andmkent** [comment]
Could you elaborate on 'library code should not be raising contract
errors'? I don't think I understand what you mean. Surely we're okay with
(+ 3 "four") raising a contract error, so I feel like there's a distinction
you're thinking of between these two cases that is important to your point.

**Comment by rfindler** [comment]
Shouldn't be raising contract errors _that blame the library_. See the
original message.

Robby

**Comment by lexi-lambda** [comment]
Well, the "quick fix" solution would be to rewrite `typed/rackunit` to use the base-env-extending module language, which wouldn't generate any additional contracts, anyway. I don't like that, especially since rackunit doesn't need to be performance-critical, but it's there.

The other option would be to implement wrapper functions on the typed side as suggested in that mailing list thread. The disadvantage to this, of course, is unnecessary duplication.

**Comment by samth** [comment]
Unfortunately, in the general case we don't have great options.

* We can give this dynamic error (perhaps with a better error message such as "opaque value passed as `Any` to untyped code").
* We can give a much more restricted type to `check-eq?`, so it statically rejects this program, which would also mean that it no longer works on functions or transparent structs or other kinds of data.

For `typed/rackunit`, we can reimplement `check-eq?` as @ntoronto's post describes. But in general, using `Any` on the interface between typed and untyped programs is bound to run into this problem.

This relates to @rfindler's paper about contracts as projections and the ordering on contracts -- `Any` is the least restrictive type, so it needs the least restrictive contract. But as that paper points out, the least restrictive contract is just one that always blames the other side. `any-wrap/c` (the translation of `Any`) is a little less restrictive than that, but you still run into these problems. 

Usually the solution is to use a more precise type than `Any`, but that isn't possible here. So I think the current solution, with a better error message, is the best choice.

**Comment by samth** [comment]
Also, there's a slightly subtle distinction between the two libraries here -- really it's `typed/rackunit` blaming `rackunit`. That's not to say that this is an entirely ok situation.

**Comment by rfindler** [comment]
It seems like there should be a better way somehow (not sure how much work
this thought entails) but since we know that the only thing that happens is
an eq? check, there should be a way to allow that. (Perhaps I'm really just
asking for more of rackunit to be implemented in TR, tho.)

**Comment by samth** [comment]
Considering `check-equal?`, what we'd need would be some form of `chaperone-something` that (a) didn't require proof that its argument was a struct, (b) allowed `equal?` to just use the underlying value (although there are potential issues w/ `prop:equal`) and (c) would error on attempts to do other things with the struct.

That chaperone would fit nicely with the JavaScript proxy semantics, but not really with the rest of the design of Racket's chaperones right now, as far as I can see.

**Comment by rfindler** [comment]
What about "check-eqal?" implemented via two chaperone-of? checks?

Robby

**Comment by jsmaniac** [comment]
I already have implemented in my personal toolkit some wrapper macros [`check-equal?:`](https://github.com/jsmaniac/phc-toolkit/blob/master/typed-rackunit.rkt#L37) and [`check-eq?:`](https://github.com/jsmaniac/phc-toolkit/blob/master/typed-rackunit.rkt#L60). They rely on `check-true` and pass to it a lambda which performs the actual check. To obtain the original error message, `with-check-info*` is used.

```
raco pkg install phc-toolkit
```

Then the following works:

```
#lang typed/racket/base

(require phc-toolkit/typed-rackunit)

(define-struct foo ([bar : Any]))
(define foo1 (make-foo 42))

(check-eq?: foo1 foo1)
```

Also, the macro supports annotating the values so that their type is checked at compile-time:

```
(check-eq?: foo1 : foo foo1)
;; is syntactic sugar for:
(check-eq?: (ann foo1 foo) foo1)
```

For now only a few `check-…` are implemented, but adding new ones is very easy. Maybe something like this could be used in `typed/rackunit`?

**Comment by samth** [comment]
Those sound like great additions to the library.
